### PR TITLE
Giving warnings on accessing fields of invalid headers

### LINKED ISSUE #2784: Compiler does not give warnings on accessing fields of invalid headers

### PR DESCRIPTION
**Pull request by kostasyrmia**
Fix #2784 
This improves the FindUninitialized pass to report warnings on accessing fields of invalid headers. Currently, the warnings are reported only for headers and header fields of structs, but not for header stacks and header unions.

According to the specification, in the following cases headers become valid:

 - calling the setValid method of a header
 - calling the extract extern method of the packet_in type
 - initialization with a list expression
 - initialization with some valid header

In the following cases headers become invalid:

 - calling the setInvalid method of a header
 - initialization with some invalid header

During visiting of the statements in the program, the current values of header valid bits are stored in a map, which is updated each time some of these cases are encountered. When encountering if and switch statements, each branch is entered with the same information that the pass had until that point, and at the end of these statements, the common part of all branches is taken.

Currently, the conditions of if statements are not processed. The pass is visiting each branch with the information that it had until that point. This can lead to giving some false warnings in the situation where there are calls to the isValid method in the condition. For example, the following snippet of code would give warnings for every access to hdr fields in the if block, even though the header hdr is guaranteed to be valid inside the if block.

    hdr.setInvalid();
    if (hdr.isValid()) {   
      // read some hdr fields here   
    }
The expression in the condition can become more complex by adding &&, || and !. So to prevent giving false warnings in this situation, each time isValid call is encountered, that header is temporarily put in a set of 'not-to-report' headers until the end of the block or until that header is again explicitly set to valid or invalid inside that block.

### CODE REVIEW

**Comment by kostasyrmia** [comment]
@mbudiu-vmw
I have fixed the issues u have requested and I have double-checked the -error files and they seem ok to me. Could you please check them too?

**Comment by kostasyrmia** [comment]
I forgot to mention the issue in the gauntlet_hdr_assign_1-bmv2.p4 test. If I am not wrong, the first FindUninitialized pass ends without warnings, then the RemoveUnused pass removes the assignment x = h.h in the ingress control, and finally, when these passes are repeated, we get some warnings, which I think should not be reported.

**Comment by kostasyrmia** [comment]
Also in addition to warnings for invalid headers, I have added warnings for potentially invalid headers because I thought that it would be useful too.  I am interested in your opinion on that. Do you think they should be reported or rather not?

**Comment by mbudiu-vmw** [comment]
> 
> 
> Also in addition to warnings for invalid headers, I have added warnings for potentially invalid headers because I thought that it would be useful too. I am interested in your opinion on that. Do you think they should be reported or rather not?

If such warnings generate too many false positives they should be avoided. People really don't like spurious warnings.

File: testdata/p4_16_samples_outputs/gauntlet_hdr_assign_1-bmv2.p4-stderr [hunk 1]
```diff
@@ -0,0 +1,6 @@
+gauntlet_hdr_assign_1-bmv2.p4(38): [--Wwarn=invalid_header] warning: accessing a field of an invalid header x
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples_outputs/gauntlet_hdr_assign_1-bmv2.p4-stderr:1]: I think this message is wrong. The previous line was `x = h.h`, so x.a may actually be valid if h.h.a was valid.

**Comment by mbudiu-vmw** [comment]
Without reading your code buy by looking at the produced error message your analysis is still producing wrong warnings.

File: testdata/p4_16_samples_outputs/gauntlet_hdr_assign_1-bmv2.p4-stderr (hunk 1 above)
Comment (kostasyrmia) [reply testdata/p4_16_samples_outputs/gauntlet_hdr_assign_1-bmv2.p4-stderr:1]: Yes, I am aware of this issue. I commented today (above) about this situation, I think that the RemoveUnused pass removed that line.

> I forgot to mention the issue in the gauntlet_hdr_assign_1-bmv2.p4 test. If I am not wrong, the first FindUninitialized pass ends without warnings, then the RemoveUnused pass removes the assignment x = h.h in the ingress control, and finally, when these passes are repeated, we get some warnings, which I think should not be reported.

Here is the final output after all passes in simplifyDefUse finish:

    apply /* 
    <BlockStatement>(74255) */
    {
        /* 
      <MethodCallStatement>(62620)
        <MethodCallExpression>(62621)
          <Member>(62622)setInvalid
            <PathExpression>(62623)
              x_0/x
          <Vector<Type>>(61096), size=0
          <Vector<Argument>>(61094), size=0 */
        x_0.setInvalid();
        /* 
        <EmptyStatement>(74270) */    <-------------- this line
        ;
        /* 
      <AssignmentStatement>(62635)
        <Member>(62636)a
          <PathExpression>(62637)
            x_0/x
        <Constant>(1313) 2
          <Type_Bits>(289) */
        x_0.a = 16w2;
        /* 
      <AssignmentStatement>(62642)
        <Member>(46198)eth_type
          <Member>(46199)eth_hdr
            <PathExpression>(46200)
              h
        <Member>(62647)a
          <PathExpression>(62648)
            x_0/x */
        h.eth_hdr.eth_type = x_0.a;
    }

If you want, I can try to find out why this is happening.

**Comment by kostasyrmia** [comment]
> > Also in addition to warnings for invalid headers, I have added warnings for potentially invalid headers because I thought that it would be useful too. I am interested in your opinion on that. Do you think they should be reported or rather not?
> 
> If such warnings generate too many false positives they should be avoided. People really don't like spurious warnings.

Well, I think it works well for these small programs, but I am not sure how it would behave in some more serious programs. In any case, we can remove it if it starts bothering us.

File: testdata/p4_16_samples_outputs/gauntlet_hdr_assign_1-bmv2.p4-stderr (hunk 1 above)
Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/gauntlet_hdr_assign_1-bmv2.p4-stderr:1]: If that's what's happening it's clearly a bug, since x is not unused. Please file an issue about it.

**Comment by mbudiu-vmw** [comment]
Let's merge first https://github.com/p4lang/p4c/pull/2891, then you can rebase and I will review the code.
This looks like a very useful analysis based on the number of warnings produced.

**Comment by kostasyrmia** [comment]
I have updated the branch, I think everything is ready for review.

File: testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-error [hunk 2]
```diff
@@ -0,0 +1,3 @@
+psa-table-hit-miss.p4(62): [--Wwarn=invalid_header] warning: accessing a field of a potentially invalid header hdr.ethernet
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-error:1]: what is this file

File: testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-stderr [hunk 3]
```diff
@@ -0,0 +1,3 @@
+psa-table-hit-miss.p4(62): [--Wwarn=invalid_header] warning: accessing a field of a potentially invalid header hdr.ethernet
+            hdr.ethernet.srcAddr : exact;
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-stderr:2]: I am not sure that this message is desirable. Why does it appear here?
Because you are assuming the fields of hdr may be invalid?

File: frontends/p4/simplifyDefUse.cpp [hunk 4]
```diff
@@ -97,6 +98,150 @@ class HasUses {
...
+    }
+    void update(const IR::Expression* expr, bool valid) {
+        if (!expr)
+            return;
+
+        auto storage = getStorageLocation(expr);
+        update(storage, valid);
+    }
+
+    const bool* find(const StorageLocation* storage) const {
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply frontends/p4/simplifyDefUse.cpp:181]: There is also an enum in removeReturns.h that has 3 values (Yes, No, Maybe).
We could move that to a separate file and name it "TernaryBool" and reuse it.

File: frontends/p4/simplifyDefUse.cpp [hunk 5]
```diff
@@ -179,6 +332,24 @@ class FindUninitialized : public Inspector {
...
+                    initHeaderParam(f, value);
+            }
+        }
+    }
+
+    // Called at the beginning of controls, parsers and functions
+    void initHeaderParams(const IR::ParameterList* parameters) {
+        for (auto p : parameters->parameters)
+            if (auto storage = definitions->storageMap->getStorage(p))
+                initHeaderParam(storage, p->direction != IR::Direction::Out);
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:350]: I guess you can't really have header constants (with IR::Direction::None), but perhaps you should exclude these too.

File: frontends/p4/simplifyDefUse.cpp [hunk 6]
```diff
@@ -271,8 +451,63 @@ class FindUninitialized : public Inspector {
...
+            FindUninitialized fu(this, currentPoint);
+            (void)state->apply(fu);
+
+            auto next = transitions.getCallees(state);
+            for (auto n : *next) {
+                if (inputHeaderDefs.find(n) == inputHeaderDefs.end()) {
+                    inputHeaderDefs[n] = headerDefs->clone();
+                    toRun.emplace(n);
+                } else {
+                    auto joined = inputHeaderDefs[n]->intersect(headerDefs);
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:491]: usually "join" means "union", so perhaps you want to rename this variable to something like 'common'.

File: frontends/p4/simplifyDefUse.cpp [hunk 7]
```diff
@@ -339,6 +574,83 @@ class FindUninitialized : public Inspector {
...
+
+            auto it1 = dst_fields.begin();
+            auto it2 = src_fields.begin();
+            while (it1 != dst_fields.end() && it2 != src_fields.end()) {
+                processHeadersInAssignment(*it1, *it2);
+                ++it1;
+                ++it2;
+            }
+        }
+    }
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:610]: BUG_CHECK to make sure that no other cases exist?

File: frontends/p4/simplifyDefUse.cpp [hunk 8]
```diff
@@ -339,6 +574,83 @@ class FindUninitialized : public Inspector {
...
+        if (auto dst_struct_storage = dst->to<StructLocation>()) {
+            if (dst_struct_storage->isHeader()) {
+                if (src->is<IR::StructExpression>()) {
+                    // always sets the valid bit to true
+                    headerDefs->update(dst, true);
+                } else if (src->is<IR::MethodCallExpression>()) {
+                    // return value of a function
+                    headerDefs->update(dst, true);
+                } else if (src->is<IR::ArrayIndex>()) {
+                    // TODO: header stacks
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:627]: Can't you just keep track of a valid bit for each slot of the array?
If the index is not constant you can assume the "or" of all valid bits of all elements for reads.
You can do this in a subsequent PR.

File: frontends/p4/simplifyDefUse.cpp [hunk 9]
```diff
@@ -558,20 +897,67 @@ class FindUninitialized : public Inspector {
...
         return false;
     }
 
+    void reportWarningIfInvalidHeader(const IR::Expression* expression, const IR::Type *expr_type) {
+        if (!reportInvalidHeaders)
+            return;
+
+        if (expr_type->is<IR::Type_Header>()) {
+            if (auto member = expression->to<IR::Member>()) {
+                // TODO: header unions
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:933]: you may be able to model a header_union as a struct of headers.
also, writing to a union field implicitly invalidates all other fields.
this can be a separate PR.

File: frontends/p4/simplifyDefUse.cpp [hunk 10]
```diff
@@ -589,18 +975,47 @@ class FindUninitialized : public Inspector {
...
+            auto am = mi->to<ApplyMethod>();
+            isControlOrParserApply = !am->isTableApply();
+        }
         for (auto p : *mi->substitution.getParametersInArgumentOrder()) {
             auto expr = mi->substitution.lookup(p);
             if (p->direction != IR::Direction::Out) {
                 visit(expr);
             }
+
+            // We assume control and parser apply calls set all output headers to valid
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:1003]: that should be true of extern functions too.

File: frontends/p4/simplifyDefUse.cpp [hunk 11]
```diff
@@ -614,6 +1029,18 @@ class FindUninitialized : public Inspector {
                 callee.push_back(table);
             }
         } else if (auto em = mi->to<ExternMethod>()) {
+            auto args = em->expr->arguments;
+            if (args->size() > 0) {
+                auto header = args->at(0);
+                if (em->originalExternType->getName().name == "packet_in" &&
+                    em->method->getName().name == "extract") {
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:1036]: if you assume that extern methods will potentially set the valid bit of every header you probably don't need this case anymore.

File: frontends/p4/simplifyDefUse.cpp [hunk 12]
```diff
@@ -614,6 +1029,18 @@ class FindUninitialized : public Inspector {
...
+                        headerDefs->update(header->expression, true);
+                } else if (em->originalExternType->getName().name == "packet_out" &&
+                         em->method->getName().name == "emit") {
+                    reportWarningIfInvalidHeader(header->expression,
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:1040]: why is this method special? all methods that have an 'in' argument of type containing a header should be handled in the same way.

**Review by mbudiu-vmw — APPROVED** [review]
I will approve, but perhaps you can still take a look at the comments.

File: testdata/p4_16_samples_outputs/gauntlet_ctrl_plane_hdr.p4-stderr [hunk 13]
```diff
@@ -0,0 +1,3 @@
+gauntlet_ctrl_plane_hdr.p4(25): [--Wwarn=invalid_header] warning: accessing a field of a potentially invalid header ctrl_hdr
```
[the remark this answers is not in the document]
Comment (kostasyrmia) [reply testdata/p4_16_samples_outputs/gauntlet_ctrl_plane_hdr.p4-stderr:1]: Making the warning function check the context itself would be the best solution in my opinion, but I am not sure how exactly it could be done. I agree about the annotation, I will check it for some of the subsequent commits.

File: testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-error (hunk 2 above)
Comment (kostasyrmia) [reply testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-error:1]: Well, ironically, when I pushed yesterday, this test failed. I checked the logs and there was a message that this output was expected in the file. This was very strange considering that I didn't change any of the files and the tests were passing before. I will remove this file and try again.

File: testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-stderr (hunk 3 above)
Comment (kostasyrmia) [reply testdata/p4_16_samples_outputs/psa-table-hit-miss.p4-stderr:2]: If I am not wrong, after the first if statement in that control, the header will be invalid after the then branch is executed, or will be valid if the condition in the if is false. So for the second apply call, the header might be invalid.
But I am considering removing the line that gives warnings about them for now (until I test the potentially invalid header analysis thoroughly).

File: frontends/p4/simplifyDefUse.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kostasyrmia) [reply frontends/p4/simplifyDefUse.cpp:181]: That looks like exactly what I needed.

File: frontends/p4/simplifyDefUse.cpp (hunk 7 above)
Comment (kostasyrmia) [reply frontends/p4/simplifyDefUse.cpp:610]: That's a great idea!

File: frontends/p4/simplifyDefUse.cpp (hunk 8 above)
Comment (kostasyrmia) [reply frontends/p4/simplifyDefUse.cpp:627]: I didn't want to put everything in a single PR because I thought it would make the review harder. I agree with your idea, it looks good to me.

File: frontends/p4/simplifyDefUse.cpp (hunk 12 above)
Comment (kostasyrmia) [reply frontends/p4/simplifyDefUse.cpp:1040]: Yes, I agree. This method isn't much more special than other extern methods.

**Comment by kostasyrmia** [comment]
I will add these suggestions, please don't merge until then.

File: testdata/p4_16_samples_outputs/gauntlet_ctrl_plane_hdr.p4-stderr (hunk 13 above)
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/gauntlet_ctrl_plane_hdr.p4-stderr:1]: I have filed #2895 and I will try to do it myself.

**Comment by kostasyrmia** [comment]
I have added a new commit be38a4b91a712a58af7890444da74db18d6d56e4 with the changes.

**Comment by kostasyrmia** [comment]
> I have added a new commit be38a4b with the changes.

@mbudiu-vmw 
I didn't tag you in the previous comment, this is in case you haven't seen it.

**Review by mbudiu-vmw — APPROVED** [review]
